[BUG] Service is unable to recover if the temporary driver-directory gets removed

**Context:**
- Playwright Version: 1.17.0
- Operating System: Linux
- Browser: All

<!-- CLI to auto-capture this info -->
<!-- npx envinfo --preset playwright --markdown -->


**Describe the bug**

Playwright uses a `/tmp` directory for creating the driver directory:

https://github.com/microsoft/playwright-java/blob/8d84afccec214856852fd812aa26a50e88e1f9e7/driver-bundle/src/main/java/com/microsoft/playwright/impl/DriverJar.java#L31-L34

However, if this directory gets removed, the service will not recover from the situation, but instead all attempts to launch a new browser will fail:

```
Caused by: java.io.IOException: Cannot run program "/tmp/playwright-java-16255647668257527045/playwright.sh": error=2, No such file or directory
```

I'm filing this as a bug because I would argue it should be safe to remove old files from the `/tmp` directory. We encountered this issue, because [tmpreaper](https://manpages.ubuntu.com/manpages/bionic/man8/tmpreaper.8.html) removed the driver directory after seven days.

I think ideally the service should create a new temporary directory if the old one gets removed. Also, it might be beneficial to be able to configure where these driver directories are created?

--- a/driver-bundle/src/main/java/com/microsoft/playwright/impl/DriverJar.java
+++ b/driver-bundle/src/main/java/com/microsoft/playwright/impl/DriverJar.java
@@ -28,8 +28,14 @@ public class DriverJar extends Driver {
   private static final String PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD = "PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD";
   private final Path driverTempDir;
 
-  DriverJar() throws IOException {
-    driverTempDir = Files.createTempDirectory("playwright-java-");
+  public DriverJar() throws IOException {
+    // Allow specifying custom path for the driver installation
+    // See https://github.com/microsoft/playwright-java/issues/728
+    String alternativeTmpdir = System.getProperty("playwright.driver.tmpdir");
+    String prefix = "playwright-java-";
+    driverTempDir = alternativeTmpdir == null
+      ? Files.createTempDirectory(prefix)
+      : Files.createTempDirectory(Paths.get(alternativeTmpdir), prefix);
     driverTempDir.toFile().deleteOnExit();
   }
 
@@ -48,10 +54,9 @@ public class DriverJar extends Driver {
       System.out.println("Skipping browsers download because `PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD` env variable is set");
       return;
     }
-    String cliFileName = super.cliFileName();
-    Path driver = driverTempDir.resolve(cliFileName);
+    Path driver = driverPath();
     if (!Files.exists(driver)) {
-      throw new RuntimeException("Failed to find " + cliFileName + " at " + driver);
+      throw new RuntimeException("Failed to find driver: " + driver);
     }
     ProcessBuilder pb = new ProcessBuilder(driver.toString(), "install");
     pb.environment().putAll(env);

--- a/driver-bundle/src/test/java/com/microsoft/playwright/TestInstall.java
+++ b/driver-bundle/src/test/java/com/microsoft/playwright/TestInstall.java
@@ -17,7 +17,9 @@
 package com.microsoft.playwright;
 
 import com.microsoft.playwright.impl.Driver;
+import com.microsoft.playwright.impl.DriverJar;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -40,5 +42,12 @@ public class TestInstall {
     Process p = pb.start();
     boolean result = p.waitFor(1, TimeUnit.MINUTES);
     assertTrue(result, "Timed out waiting for browsers to install");
+  }
+
+  @Test
+  void playwrightDriverInAlternativeTmpdir(@TempDir Path tmpdir) throws Exception {
+    System.setProperty("playwright.driver.tmpdir", tmpdir.toString());
+    DriverJar driver = new DriverJar();
+    assertTrue(driver.driverPath().startsWith(tmpdir), "Driver path: " + driver.driverPath() + " tmp: " + tmpdir);
   }
 }

--- a/driver-bundle/src/test/java/com/microsoft/playwright/TestInstall.java
+++ b/driver-bundle/src/test/java/com/microsoft/playwright/TestInstall.java
@@ -18,6 +18,8 @@ package com.microsoft.playwright;
 
 import com.microsoft.playwright.impl.Driver;
 import com.microsoft.playwright.impl.DriverJar;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
@@ -29,10 +31,15 @@ import java.util.concurrent.TimeUnit;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public class TestInstall {
-  @Test
-  void playwrightCliInstalled() throws Exception {
+  @BeforeEach
+  void clearSystemProperties() {
     // Clear system property to ensure that the driver is loaded from jar.
     System.clearProperty("playwright.cli.dir");
+    System.clearProperty("playwright.driver.tmpdir");
+  }
+
+  @Test
+  void playwrightCliInstalled() throws Exception {
     Path cli = Driver.ensureDriverInstalled(Collections.emptyMap());
     assertTrue(Files.exists(cli));
 

--- a/driver/src/main/java/com/microsoft/playwright/impl/Driver.java
+++ b/driver/src/main/java/com/microsoft/playwright/impl/Driver.java
@@ -54,15 +54,15 @@ public abstract class Driver {
         throw new RuntimeException("Failed to create driver", exception);
       }
     }
-    String name = instance.cliFileName();
-    return instance.driverDir().resolve(name);
+    return instance.driverPath();
   }
 
   protected abstract void initialize(Map<String, String> env) throws Exception;
 
-  protected String cliFileName() {
-    return System.getProperty("os.name").toLowerCase().contains("windows") ?
+  public Path driverPath() {
+    String cliFileName = System.getProperty("os.name").toLowerCase().contains("windows") ?
       "playwright.cmd" : "playwright.sh";
+    return driverDir().resolve(cliFileName);
   }
 
   private static Driver createDriver() throws Exception {

